rclcpp/include/rclcpp/node.hpp

```diff
@@ -139,7 +139,7 @@ class Node : public std::enable_shared_from_this<Node>
   const std::vector<rclcpp::callback_group::CallbackGroup::WeakPtr> &
   get_callback_groups() const;
 
-  /// Create and return a Publisher.
+  /// Create and return a Publisher. Note: This constructor is deprecated
```

Comment:
Does the [deprecate attribute](https://en.cppreference.com/w/cpp/language/attributes/deprecated) work for constructors?

Reply:
The attribute should work (my comment may be misnaming, this are factory functions, not contsructors) - My one concern with marking these as deprecated is that it will throw warnings in other parts of the core until they are migrated, which will cause the CI builds to be marked as unstable - I'm not sure if we want to do that yet.

Reply:
> My one concern with marking these as deprecated is that it will throw warnings in other parts of the core until they are migrated, which will cause the CI builds to be marked as unstable - I'm not sure if we want to do that yet.

Commenting that they are deprecated without deprecating them using the attribute is definitely not what we want to do, it should either be removed, or deprecated with the attribute (with or without a comment in the doc block), or not deprecated at all.

I think we should deprecate it with the attribute, and look at the downstream impact (how many places need to be updated). If the number is low, then we can consider just removing it directly.

@emersonknapp can you please collect a list of all the places that need to be touched in code listed in the `ros2.repos` file? You could do this by deprecating this interface and running CI, which will collect the compiler warnings for you in one place.

As we also spoke about off-line, I do think we should address the issue of removing or deprecating the existing interfaces and fixing any new downstream failures or warnings due to that all at once, rather than a two stage process where we merge the new API and then separately get rid of the other ones.

Reply:
This applies to all the interfaces here that may need to be deprecated.

Reply:
Good plan, I will go ahead and make those changes and run CI